File: app/src/main/java/cn/ifafu/ifafu/widget/SyllabusWidget.java
package cn.ifafu.ifafu.widget;

import android.annotation.SuppressLint;
import android.app.PendingIntent;
import android.appwidget.AppWidgetManager;
import android.appwidget.AppWidgetProvider;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.view.View;
import android.widget.RemoteViews;

import java.text.MessageFormat;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

import cn.ifafu.ifafu.R;
import cn.ifafu.ifafu.app.Constant;
import cn.ifafu.ifafu.app.IFAFU;
import cn.ifafu.ifafu.data.entity.NextCourse;
import cn.ifafu.ifafu.mvp.other.SplashActivity;
import cn.ifafu.ifafu.mvp.syllabus.SyllabusActivity;
import cn.ifafu.ifafu.mvp.syllabus.SyllabusModel;

public class SyllabusWidget extends AppWidgetProvider {

    private static String ACTION_WIDGET_SYLLABUS_CLICK = "ifafu.widget.syllabus.REFRESH";

    private void updateSyllabusWidget(Context context, AppWidgetManager appWidgetManager,
                                      int appWidgetId) {
        RemoteViews views = new RemoteViews(context.getPackageName(), R.layout.syllabus_widget);
        updateSyllabusWidget(context, views);
        appWidgetManager.updateAppWidget(appWidgetId, views);
    }

    private PendingIntent getPendingIntent(Context context, int resId) {
        Intent intent = new Intent();
        intent.setClass(context, SyllabusWidget.class);
        intent.setData(Uri.parse("Id:" + resId));
        intent.setAction(ACTION_WIDGET_SYLLABUS_CLICK);
        return PendingIntent.getBroadcast(context, 0, intent, 0);
    }

    @SuppressLint("DefaultLocale")
    private void updateSyllabusWidget(Context context, RemoteViews remoteViews) {

        remoteViews.setOnClickPendingIntent(R.id.tv_refresh,
                getPendingIntent(context, R.id.tv_refresh));
        remoteViews.setOnClickPendingIntent(R.id.btn_jump,
                getPendingIntent(context, R.id.btn_jump));

        SimpleDateFormat format = new SimpleDateFormat("HH:mm:ss", Locale.CHINA);
        SimpleDateFormat format1 = new SimpleDateFormat("MM月dd日 E", Locale.CHINA);
        Date date = new Date();
        remoteViews.setTextViewText(R.id.tv_refresh_time, context.getString(R.string.refresh_time_format, format.format(date)));

        SyllabusModel model = new SyllabusModel(context);
        NextCourse next = model.getNextCourse();
        switch (next.getResult()) {
            case NextCourse.IN_HOLIDAY:
                remoteViews.setTextViewText(R.id.tv_time, String.format("%s 放假中", format1.format(date)));
            case NextCourse.EMPTY_DATA:
            case NextCourse.NO_TODAY_COURSE:
            case NextCourse.NO_NEXT_COURSE:
                isShowCourseInfo(remoteViews, false);
                remoteViews.setTextViewText(R.id.tv_tip, next.getTitle());
                break;
            case NextCourse.HAS_NEXT_COURSE:
                isShowCourseInfo(remoteViews, true);
                remoteViews.setTextViewText(R.id.tv_course_name, next.getTitle() + next.getName());
                remoteViews.setTextViewText(R.id.tv_course_address, next.getAddress());
                remoteViews.setTextViewText(R.id.tv_course_node,
                        MessageFormat.format("{0} {1}", next.getNodeText(), next.getTimeText()));
                break;
        }
        remoteViews.setTextViewText(R.id.tv_time, String.format("%s %s", format1.format(date), next.getWeekText()));
    }

    private void isShowCourseInfo(RemoteViews remoteViews, boolean isShow) {
        if (isShow) {
            remoteViews.setViewVisibility(R.id.tv_course_name, View.VISIBLE);
            remoteViews.setViewVisibility(R.id.tv_course_node, View.VISIBLE);
            remoteViews.setViewVisibility(R.id.tv_course_address, View.VISIBLE);
            remoteViews.setViewVisibility(R.id.tv_tip, View.GONE);
        } else {
            remoteViews.setViewVisibility(R.id.tv_course_name, View.GONE);
            remoteViews.setViewVisibility(R.id.tv_course_node, View.GONE);
            remoteViews.setViewVisibility(R.id.tv_course_address, View.GONE);
            remoteViews.setViewVisibility(R.id.tv_tip, View.VISIBLE);
        }
    }

    @Override
    public void onUpdate(Context context, AppWidgetManager appWidgetManager, int[] appWidgetIds) {
        for (int appWidgetId : appWidgetIds) {
            updateSyllabusWidget(context, appWidgetManager, appWidgetId);
        }
    }

    @Override
    public void onReceive(Context context, Intent intent) {
        if (intent.getAction() == null) {
            return;
        }
        if (intent.getAction().equals(ACTION_WIDGET_SYLLABUS_CLICK)) {
            RemoteViews remoteViews = new RemoteViews(context.getPackageName(), R.layout.syllabus_widget);
            Uri data = intent.getData();
            int resId = -1;
            if (data != null) {
                resId = Integer.parseInt(data.getSchemeSpecificPart());
            }
            switch (resId) {
                case R.id.tv_refresh:
                    updateSyllabusWidget(context, remoteViews);
                    break;
                case R.id.btn_jump:
                    Intent jumpIntent;
                    if (IFAFU.FIRST_START_APP) {
                        jumpIntent = new Intent(context, SplashActivity.class);
                        jumpIntent.putExtra("jump", Constant.SYLLABUS_ACTIVITY);
                    } else {
                        jumpIntent = new Intent(context, SyllabusActivity.class);
                    }
                    jumpIntent.putExtra("from", Constant.SYLLABUS_WIDGET);
                    jumpIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                    context.startActivity(jumpIntent);
                    break;
            }
            AppWidgetManager manger = AppWidgetManager.getInstance(context);
            ComponentName thisName = new ComponentName(context, SyllabusWidget.class);
            manger.updateAppWidget(thisName,remoteViews);
        } else {
            super.onReceive(context, intent);
        }
    }

}

File: app/src/main/java/cn/ifafu/ifafu/widget/SyllabusWidget.kt
package cn.ifafu.ifafu.widget

import android.app.PendingIntent
import android.appwidget.AppWidgetManager
import android.appwidget.AppWidgetProvider
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.view.View
import android.widget.RemoteViews
import cn.ifafu.ifafu.R
import cn.ifafu.ifafu.app.Constant
import cn.ifafu.ifafu.data.repository.impl.RepositoryImpl
import cn.ifafu.ifafu.ui.main.bean.ClassPreview
import cn.ifafu.ifafu.ui.schedule.SyllabusActivity
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.withContext
import java.text.SimpleDateFormat
import java.util.*

class SyllabusWidget : AppWidgetProvider() {

    companion object;

    private suspend fun updateSyllabusWidget(context: Context, remoteViews: RemoteViews) = withContext(Dispatchers.IO) {
        //设置跳转课表按钮监听
        val syllabusIntent: PendingIntent = Intent(context, SyllabusActivity::class.java)
                .let { intent ->
                    intent.putExtra("from", Constant.SYLLABUS_WIDGET)
                    PendingIntent.getActivity(context, 0, intent, 0)
                }
        remoteViews.setOnClickPendingIntent(R.id.btn_go, syllabusIntent)
        //设置刷新Widget按钮监听
        val refreshIntent: PendingIntent = Intent()
                .let { intent ->
                    intent.action = AppWidgetManager.ACTION_APPWIDGET_UPDATE
                    PendingIntent.getBroadcast(context, 0, intent, 0)
                }
        remoteViews.setOnClickPendingIntent(R.id.btn_refresh, refreshIntent)
        //更新刷新时间
        val format = SimpleDateFormat("HH:mm:ss", Locale.CHINA)
        val now = context.getString(R.string.refresh_time_format, format.format(Date()))
        remoteViews.setTextViewText(R.id.tv_refresh_time, now)
        val repo = RepositoryImpl
        try {
            val courses = repo.syllabus.getAll()
            val setting = repo.syllabus.getSetting()
            val holidayFromToMap = repo.syllabus.getAdjustmentInfo()
            val preview = ClassPreview.convert(courses, holidayFromToMap, setting)
            setPreviewViewInfo(context, remoteViews, preview)
        } catch (e: Exception) {
            remoteViews.setViewVisibility(R.id.message, View.VISIBLE)
            remoteViews.setViewVisibility(R.id.layout_info, View.GONE)
            remoteViews.setTextViewText(R.id.message, e.message)
        }
    }

    override fun onUpdate(context: Context, appWidgetManager: AppWidgetManager, appWidgetIds: IntArray) {
        appWidgetIds.forEach { appWidgetId ->
            val views = RemoteViews(context.packageName, R.layout.widget_syllabus)
            runBlocking {
                updateSyllabusWidget(context, views)
            }
            appWidgetManager.updateAppWidget(appWidgetId, views)
        }
    }

    override fun onReceive(context: Context, intent: Intent) {
        if (intent.action == AppWidgetManager.ACTION_APPWIDGET_UPDATE) {
            val remoteViews = RemoteViews(context.packageName, R.layout.widget_syllabus)
            runBlocking {
                updateSyllabusWidget(context, remoteViews)
            }
            val manger = AppWidgetManager.getInstance(context)
            val provider = ComponentName(context, SyllabusWidget::class.java)
            //更新Widget
            manger.updateAppWidget(provider, remoteViews)
        } else {
            super.onReceive(context, intent)
        }
    }

    private fun setPreviewViewInfo(context: Context, remoteViews: RemoteViews, preview: ClassPreview) {
        remoteViews.setTextViewText(R.id.tv_week_time, preview.dateText)
        if (preview.hasInfo) {
            with(remoteViews) {
                //显示详情信息
                setViewVisibility(R.id.message, View.GONE)
                setViewVisibility(R.id.layout_info, View.VISIBLE)
                setTextViewText(R.id.location, preview.address)
                setTextViewText(R.id.classTime, preview.classTime)
                setTextViewText(R.id.timeLeft, preview.timeLeft)
                val numOfClass = preview.numberOfClasses
                setTextViewText(R.id.tv_total, context.getString(R.string.node_format, numOfClass[0], numOfClass[1]))
                if (preview.isInClass) {
                    setTextViewText(R.id.tv_next, context.getString(R.string.now_class_format, preview.nextClass))
                    setTextViewText(R.id.tv_status, context.getString(R.string.in_class))
                    setTextViewCompoundDrawables(R.id.tv_status, R.drawable.ic_point_blue, 0, 0, 0)
                } else {
                    setTextViewText(R.id.tv_next, context.getString(R.string.next_class_format, preview.nextClass))
                    setTextViewText(R.id.tv_status, context.getString(R.string.out_class))
                    setTextViewCompoundDrawables(R.id.tv_status, R.drawable.ic_point_red, 0, 0, 0)
                }
            }
        } else {
            remoteViews.setViewVisibility(R.id.message, View.VISIBLE)
            remoteViews.setViewVisibility(R.id.layout_info, View.GONE)
            remoteViews.setTextViewText(R.id.message, preview.message)
        }
    }
}

